Restructure documentation around guides (#107)
Qiskit/qiskit-fermions: docs/guides/operators.rst

.. _operators_explanation:

Design Principles of Operator Representations
==============================================

This guide explains the common design principles and core concepts shared across
all operator representations in the :mod:`~qiskit_fermions.operators` module.

Overview
--------

The operator representations provided by this module share several fundamental
design principles:

> - **Sparse Data Structure** (`read more <sparse_term_representation_>`_):

Reviewer: ```suggestion
- **`Sparse data structure <sparse_term_representation_>`_**:
```

  Operators encode only non-identity operations. The internal data layout is
  generally inspired by sparse matrix data formats, enabling efficient storage
  and computation for systems with many modes but relatively few significant
  contributions.

> - **Term Iteration and Reconstruction** (`read more <term_iteration_and_reconstruction_>`_):

Reviewer: ```suggestion
- **`Term iteration and reconstruction <term_iteration_and_reconstruction_>`_**:
```

  Despite internal sparse storage, operators provide a consistent iteration interface
  that lets you inspect, filter, and transform terms without understanding the
  underlying data structure, then reconstruct new operators from modified terms.

> - **Mode-based Indexing** (`read more <mode_based_indexing_>`_):

Reviewer: ```suggestion
- **`Mode-based indexing <mode_based_indexing_>`_**:
```

  Operators use abstract mode indices to label fermionic degrees of freedom,
  enabling flexible mapping from physical systems to the operator representation.

> - **Built-in Term Grouping** (`read more <term_grouping_>`_):

Reviewer: ```suggestion
- **`Built-in term grouping <term_grouping_>`_**:
```

  Operators natively support grouping information that associates terms with
  group indices. This enables optimizations and physical structure preservation

>   without requiring separate data structures. See :ref:`the grouping explanation

Reviewer: ```suggestion
  without requiring separate data structures. See the :ref:`grouping 
```

  guide <grouping_explanation>` for practical usage.

> - **Rich Arithmetic Interface** (`read more <arithmetic_and_mathematical_operations_>`_):

Reviewer: ```suggestion
- **`Rich arithmetic interface <arithmetic_and_mathematical_operations_>`_**:
```

  All operators implement a consistent set of arithmetic operations (addition,

>   multiplication, composition) and mathematical functions via the

Reviewer: ```suggestion
  multiplication, and composition) and mathematical functions by using the
```

  :class:`.OperatorTrait` protocol, enabling uniform code across different
  operator types.

> - **Term Ordering and Normal Forms** (`read more <operator_term_ordering_>`_):

Reviewer: ```suggestion
- **`Term ordering and normal Formse <operator_term_ordering_>`_**:
```

  Mathematically equivalent operators can have very different representations and
  behavior in quantum algorithms. All operator representations support various
  normal forms (based on algebra-specific commutation relations) to achieve
  canonical, predictable, and optimizable operator representations.

.. _sparse_term_representation:

Sparse data structure
---------------------

All operators use a **sparse representation** where only non-identity operations
are tracked. Each non-identity operation consists of a **coefficient** (complex
number) and a **sequence of actions** on specific modes.

This approach dramatically reduces memory usage and computation time, especially for
systems with many modes but relatively few significant contributions. By encoding only
non-identity operations, operations focus only on what matters, enabling work with large
systems that would be infeasible with dense representations. Additionally, operators

> naturally scale to any number of modes—an operator acting on modes ``{0, 1}`` works

Reviewer: ```suggestion
naturally scale to any number of modes - an operator acting on modes ``{0, 1}`` works
```

unchanged in systems with many more modes since unaffected modes are implicitly identity.

Note that identical terms are preserved separately during arithmetic operations
and must be explicitly combined if needed.

Internal storage format
~~~~~~~~~~~~~~~~~~~~~~~

> Internally, operators are stored using arrays inspired by sparse matrix data formats:

Reviewer: ```suggestion
Internally, operators are stored in arrays inspired by sparse matrix data formats:
```

- **Coefficients array**: The complex coefficient for each term
- **Mode indices array**: The fermionic modes that each action acts upon

> - **Boundaries array**: Indices marking where each term's modes begin and end in the modes array

Reviewer: ```suggestion
- **Boundaries array**: Indices marking where each term's modes begin and end in the mode array
```

.. important::

>    Additional arrays may be present depending on the operator type. For example,

Reviewer: ```suggestion
   Additional arrays might be present depending on the operator type. For example,
```

   :class:`.FermionOperator` instances include an **actions array** of booleans
   that specifies the type of fermionic action acting on the respective mode
   index. Majorana operators do not require this distinction. See the API
   documentation for your specific operator type to understand the full storage
   format.

> The following examples show how these arrays are organized. First, here's a direct

Reviewer: ```suggestion
The following examples show how these arrays are organized. First is a direct
```

       >>>
       >>> # Combine the sparse operators
       >>> op1 += op2
       >>> print(op1)
         1.000000e0 +0.000000e0j * (+0 -1)
         1.000000e0 +0.000000e0j * (+2 -3)

    .. code-block:: c

       #include <qiskit_fermions.h>

       // Construct first operator: 1.0 * c_0 a_1
       QkComplex67 coeff1[1] = {{1.0, 0.0}};
       uint32_t modes1[2] = {0, 1};
       uint32_t boundaries1[2] = {0, 2};
       QfFermionOperator *op1 = qf_ferm_op_new(1, 2, coeff1, modes1, boundaries1);

       // Construct second operator: 1.0 * c_2 a_3
       QkComplex67 coeff2[1] = {{1.0, 0.0}};
       uint32_t modes2[2] = {2, 3};
       uint32_t boundaries2[2] = {0, 2};
       QfFermionOperator *op2 = qf_ferm_op_new(1, 2, coeff2, modes2, boundaries2);

       // Add operators
       qf_ferm_op_add_assign(op1, op2);

       qf_ferm_op_free(op1);
       qf_ferm_op_free(op2);

Convenient construction methods
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For Python developers, several convenient construction methods are available that
abstract away the sparse storage details. These make it easier to build operators
without worrying about managing coefficient, mode, and boundary arrays:

.. hint::

>    Individual operator implementations may support additional construction methods

Reviewer: ```suggestion
   Individual operator implementations might support additional construction methods
```

   suited to their specific use case. Check the API documentation for your operator
   type to see all available construction options.

.. tab-set-code::

    .. code-block:: python

       >>> from qiskit_fermions.operators import FermionOperator, cre, ann
       >>>
       >>> # Construct operators using operator algebra notation
       >>> op1 = FermionOperator.from_dict({(cre(0), ann(1)): 1.0})
       >>> op2 = FermionOperator.from_dict({(cre(2), ann(3)): 1.0})
       >>>
       >>> # The result is sparse even when combining them
       >>> op1 += op2
       >>> print(op1)
       1.000000e0 +0.000000e0j * (+0 -1)
       1.000000e0 +0.000000e0j * (+2 -3)

    .. code-block:: c

       // The C API uses direct array construction; convenience methods are not available.

.. _term_iteration_and_reconstruction:

Term iteration and reconstruction
----------------------------------

> Operators provide a consistent iteration interface via :meth:`.OperatorTrait.iter_terms`

Reviewer: ```suggestion
Operators provide a consistent iteration interface by using :meth:`.OperatorTrait.iter_terms`
```

despite their internal sparse representation. This allows you to inspect, filter,
or transform terms without needing to understand the underlying data structure.

> You can then reconstruct a new operator from the transformed terms using

Reviewer: ```suggestion
You can then reconstruct a new operator from the transformed terms by using
```

       // See the Python API documentation for :meth:`.OperatorTrait.iter_terms` and
       // :meth:`.OperatorTrait.from_terms` for equivalent functionality.

.. _mode_based_indexing:

Mode-based indexing
-------------------

All operator representations refer to the indices that their terms act upon as
**modes**. A mode is simply an index identifying a fermionic degree of freedom
in your system. The mapping from physical degrees of freedom (such as spatial
orbitals, spin states, or other quantum numbers) to mode indices is left to the
user, enabling maximum flexibility.

This abstraction is also present in the :mod:`qiskit_fermions.circuit` module,
where the :class:`.FermionicCircuit` operates on a register of fermionic modes.
In both the operator and circuit representations, modes provide a consistent,
algebra-independent way to specify which degrees of freedom participate in a
given operation.

.. important::

   **Current implementations use spinless modes**: All operator representations
   currently provided by this module treat modes as spinless fermionic degrees
   of freedom. This means if your system has both spin-up and spin-down electrons

>    or fermions, you must explicitly map them to distinct modes (e.g., modes 0-3

Reviewer: ```suggestion
   or fermions, you must explicitly map them to distinct modes (for example, modes 0-3
```

   for spin-up of 4 spatial orbitals, modes 4-7 for spin-down, or any other
   convention you choose).

   This design keeps the core representations simple and general while avoiding
   imposing a specific spin-ordering convention. Utility modules like
   :mod:`qiskit_fermions.operators.library` provide convenience functions

>    (e.g., :meth:`.FCIDump.from_file`) that handle such mappings for you when

Reviewer: ```suggestion
   (for example, :meth:`.FCIDump.from_file`) that handle such mappings for you when
```

   loading electronic structure data.

.. hint::

>    As the package evolves, spinful operator representations may be added that

Reviewer: ```suggestion
   As the package evolves, spinful operator representations might be added that
```

   natively support spin degrees of freedom in their data model. These will be
   clearly distinguished from the current spinless implementations and will
   coexist with them in the module.

.. _term_grouping:

Term grouping and commutation relationships
-------------------------------------------

Like the coefficients and mode indices, operators can optionally store a **groups array**
that associates each term with a group index. By integrating grouping directly into the
operator representation as part of the sparse data structure, grouping information
naturally travels with the operator through transformations. This enables systematic

> exploitation of structure—whether from physical properties, algebraic relationships, or

Reviewer: ```suggestion
exploitation of structure - whether from physical properties, algebraic relationships, or
```

problem-specific symmetries. The structured information can then be used in downstream

> operations like circuit synthesis and decomposition, via methods like

Reviewer: ```suggestion
operations like circuit synthesis and decomposition by using methods like
```

:meth:`.OperatorTrait.split_out_groups`.

For detailed guidance on how to group operator terms in your workflows, see

> :ref:`the grouping explanation guide <grouping_explanation>`.

Reviewer: ```suggestion
the :ref:`grouping <grouping_explanation>` guide.
```

       >>> op = MajoranaOperator.from_dict({
       ...     (gamma(0, False),): 1.0,
       ...     (gamma(1, False),): 1.0,
       ...     (gamma(2, False), gamma(3, False)): 1.0
       ... })
       >>> # Assign group indices to terms
       >>> op.groups = [0, 0, 1]
       >>> # Partition operator by groups
       >>> grouped_ops = op.split_out_groups()

    .. code-block:: c

       #include <qiskit_fermions.h>

       // Create operator with 3 terms
       QkComplex67 coeffs[3] = {{1.0, 0.0}, {1.0, 0.0}, {1.0, 0.0}};
       uint32_t modes[4] = {0, 1, 2, 3};
       uint32_t boundaries[4] = {0, 1, 2, 4};
       QfMajoranaOperator *op = qf_maj_op_new(3, 4, coeffs, modes, boundaries);

       // Assign grouping information
       uint32_t groups[3] = {0, 0, 1};
       qf_maj_op_set_groups(op, groups, 3);

       // Partition operator by groups
       QfMajoranaOperator **grouped_ops = qf_maj_op_split_out_groups(op, &num_groups);

.. _arithmetic_and_mathematical_operations:

Arithmetic and mathematical operations
---------------------------------------

All operators implement the :class:`.OperatorTrait` protocol, which provides

> a unified set of operations across different operator types. This ensures code

Reviewer: ```suggestion
a unified set of operations across different operator types. This ensures that code
```

       QfFermionOperator *difference = qf_ferm_op_sub(op, adjoint);
       QfFermionOperator *normal_ordered = qf_ferm_op_normal_ordered(difference);
       qf_ferm_op_ichop(normal_ordered, 1e-10);

       QfFermionOperator *zero = qf_ferm_op_zero();
       bool is_hermitian = qf_ferm_op_equiv(normal_ordered, zero, 1e-10);
       printf("Operator is Hermitian: %s\n", is_hermitian ? "true" : "false");

       // Clean up
       qf_ferm_op_free(op);
       qf_ferm_op_free(adjoint);
       qf_ferm_op_free(difference);
       qf_ferm_op_free(normal_ordered);
       qf_ferm_op_free(zero);

.. important::

   The example uses ``atol=1e-10`` in both ``simplify()`` and ``equiv()``.
   The ``atol`` (absolute tolerance) parameter specifies a threshold: coefficients
   with magnitude smaller than ``atol`` are treated as zero and discarded. This is
   essential for numerical stability when comparing operators, since floating-point
   arithmetic can introduce small rounding errors that would otherwise prevent
   equivalent operators from being recognized as such.

.. hint::

   While the :class:`.OperatorTrait` protocol provides a common interface,

>    individual operator implementations may offer additional convenience methods

Reviewer: ```suggestion
   individual operator implementations might offer additional convenience methods
```

   the API documentation for your specific operator type to discover all
   available functionality.

.. _operator_term_ordering:

Operator term ordering and normal forms
---------------------------------------

A fundamental challenge in quantum operator algebra is that mathematically
equivalent operators can be represented in many different ways, each with
different implications for quantum algorithms. The same operator can be written
in algebraically equivalent forms - for example, :math:`a^\dagger b` can be
expressed as :math:`ab + [a^\dagger,b]` - yet these representations lead to
different behavior in circuit synthesis, simplification, and numerical
algorithms.

The operator representations support **normal ordering** operations that
transform operators into algebra-specific canonical forms using commutation
relations. This enables reliable comparisons (two equivalent operators have
identical normal-ordered forms), reveals simplifications (commutation relations
cause terms to cancel or combine), and supports algorithms that require specific
operator forms for correctness and efficiency.

.. important::

>    Different operator representations may implement normal ordering based on

Reviewer: ```suggestion
   Different operator representations might implement normal ordering based on
```
